src/aktualizr_primary/secondary.cc

```diff
@@ -141,20 +147,53 @@ class SecondaryWaiter {
   std::unordered_set<std::string> secondaries_to_wait_for_;
 };
 
-static Secondaries createIPSecondaries(const IPSecondariesConfig& config) {
+static Secondaries createIPSecondaries(const IPSecondariesConfig& config, bool provision, Aktualizr& aktualizr) {
   Secondaries result;
-  SecondaryWaiter sec_waiter{config.secondaries_wait_port, config.secondaries_timeout_s, result};
+  if (provision) {
+    SecondaryWaiter sec_waiter{config.secondaries_wait_port, config.secondaries_timeout_s, result};
 
-  for (auto& ip_sec_cfg : config.secondaries_cfg) {
-    auto secondary = Uptane::IpUptaneSecondary::connectAndCreate(ip_sec_cfg.ip, ip_sec_cfg.port);
-    if (secondary) {
-      result.push_back(secondary);
-    } else {
-      sec_waiter.addSecondary(ip_sec_cfg.ip, ip_sec_cfg.port);
+    for (auto& ip_sec_cfg : config.secondaries_cfg) {
+      auto secondary = Uptane::IpUptaneSecondary::connectAndCreate(ip_sec_cfg.ip, ip_sec_cfg.port);
+      if (secondary) {
+        result.push_back(secondary);
+      } else {
+        sec_waiter.addSecondary(ip_sec_cfg.ip, ip_sec_cfg.port);
+      }
+    }
+
+    sec_waiter.wait();
+  } else {
+    auto secondaries_info = aktualizr.GetRegisteredSecondaries();
+
+    // for backward compatibility, we take all secondaries if their number
+    // matches, without matching on their type
+    if (secondaries_info.size() != config.secondaries_cfg.size()) {
+      secondaries_info.erase(
+          std::remove_if(secondaries_info.begin(), secondaries_info.end(),
+                         [](const SecondaryInfo& info) -> bool { return info.type != IPSecondariesConfig::Type; }));
+    }
+
+    if (secondaries_info.size() != config.secondaries_cfg.size()) {
+      std::stringstream err_ss;
+      err_ss << "Expected " << config.secondaries_cfg.size() << " IP secondaries, but device was provisioned with "
+             << secondaries_info.size();
+      throw std::runtime_error(err_ss.str());
+    }
+
+    // note: order in configuration and storage should be the same
```

Comment:
perhaps, IP and port could be stored in the DB too, so an order that secondaries are listed in the conf won't matter?

Reply:
In the same vein as the previous comment, it is made harder by the fact we've abstracted the ip secondary concept from SotaUptaneClient and the storage.

One version I've explored was to add another free form string column to the `ecu_serials` table and `SecondaryInfo` that might have to be re-parsed.

Reply:
We'll want to avoid tables with secondary type-specific columns, so the free-form string might be the best option. Tedious but might be the most future-proof.

Reply:
I agree that adding secondary specific columns into a generic table (ecu_serials here) is a bad idea. I see a few options here:

1) extend the existing ecu_serials table with an opaque column (e.g. sec_data, sec_info, etc) intended for serialized secondary specific data required for its instantiation;

2) don't alter the existing ecu_serials table. Introduce another table `ecu` with the generic fields + the column containing serialized secondary specific data;

3) Introduce another table ip_secondary that contains the ip secondary specific info. This table is managed solely by the IP secondary factory object/method/function. In this case we can avoid dependency on Aktualizr object at all, what is needed just a path to the DB if we would like to store this table in the same DB file.

4) Introduce two new tables ECU containing generic columns + the ip_secondary specific table.